keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -391,14 +391,10 @@ and more details can be found in the section
 Sidecar containers will not block Pod completion - if all regular containers
 complete, sidecar containers will be terminated.
 
-During the sidecar startup stage the restart behavior will be similar to init
-containers. If Pod `restartPolicy` is `Never`, sidecar container failed during
-startup will NOT be restarted and the whole Pod will fail. If Pod
-`restartPolicy` is `Always` or `OnFailure`, it will be restarted.
-
-Once sidecar container is started (`postStart` completed and startup probe
-succeeded), this containers will be restarted even when the Pod `restartPolicy`
-is `Never` or `OnFailure`. 
+The `restartPolicy` field for individual init containers can override the
```

Comment:
> sidecar container failed during startup will NOT be restarted

failed meant:
- PostStart failed
- Image pull failed
- CRI failed

In all these cases the implemented behavior is that we will restart the sidecar container even if the Pod restartPolicy is `Never`.

If one want to express the semantic of "fail if the sidecar cannot start" - it can be implemented by moving this logic into the Init container executed just before the sidecar. Like if certificates are needed, but we cannot download them.

@thockin any objections with this behavior?

cc: @kubernetes/api-reviewers

Reply:
One could argue that "CRI failed" and "Image pull failed" are not failures of the container, but failures to START the container, and so retrying is not a "restart" at all.

Restarting when PostStart fails is kind of dubious, though.  At that point, the container HAS started.  If the policy is "Never", it seems like we should not restart.  Tell me why I am wrong?

Reply:
> Restarting when PostStart fails is kind of dubious, though. At that point, the container HAS started. If the policy is "Never", it seems like we should not restart. Tell me why I am wrong?

We are talking about sidecar containers. They will restart if failed even in Pod with restartPolicy=Never. So the only question is whether we should allow sidecars (which has started already, just executing the PostStart) to be able to "fail" the Pod if PostStart has failed. Originally we thought it may be useful to be able to express this. But now I am not that confident. It feels a bit strange to have different restart behavior if code failed during `PostStart` or `PostStart` itself has failed.

Reply:
Sorry, I was thinking it was if the container had restart policy.  Yes, I agree with you.  They can't fail the pod that way.

Reply:
thank you for confirming.

/lgtm
/approve

Reply:
@mrunalp for actual approval